lib/liquid/traversal.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+module Liquid
+  class Traversal
+    def self.for(node, callbacks = Hash.new(proc {}))
+      kase = CASES.find { |(klass, _)| node.is_a?(klass) }&.last
+      (kase || self).new(node, callbacks)
+    end
+
+    def initialize(node, callbacks)
+      @node = node
+      @callbacks = callbacks
+    end
+
+    def callback_for(*classes, &block)
+      cb = block
+      cb = ->(node, _) { block[node] } if block.arity.abs == 1
+      cb = ->(_, _) { block[] } if block.arity.zero?
+      classes.each { |klass| @callbacks[klass] = cb }
+      self
+    end
+
+    def traverse(context = nil)
+      children.map do |node|
+        item, new_context = @callbacks[node.class][node, context]
+        [
+          item,
+          Traversal.for(node, @callbacks).traverse(
+            new_context.nil? ? context : new_context
+          )
+        ]
+      end
+    end
+
+    protected
+
+    def children
+      @node.respond_to?(:nodelist) ? Array(@node.nodelist) : []
+    end
+
+    class Assign < Traversal
+      def children
+        [@node.from]
+      end
+    end
+
+    class Case < Traversal
+      def children
+        [@node.left] + @node.blocks
+      end
+    end
+
+    class Condition < Traversal
+      def children
+        [
+          @node.left, @node.right,
+          @node.child_condition, @node.attachment
+        ].compact
+      end
+    end
+
+    class Cycle < Traversal
+      def children
+        Array(@node.variables)
+      end
+    end
+
+    class For < Traversal
+      def children
+        (super + [@node.limit, @node.from, @node.collection_name]).compact
+      end
+    end
+
+    class If < Traversal
+      def children
+        @node.blocks
+      end
+    end
+
+    class Include < Traversal
+      def children
+        [
+          @node.template_name_expr,
+          @node.variable_name_expr
+        ] + @node.attributes.values
+      end
+    end
+
+    class TableRow < Traversal
+      def children
+        super + @node.attributes.values + [@node.collection_name]
+      end
+    end
+
+    class Variable < Traversal
+      def children
+        [@node.name] + @node.filters.flatten
+      end
+    end
+
+    class VariableLookup < Traversal
+      def children
+        @node.lookups
+      end
+    end
```

Comment:
I'm not sure this is better than colocating these definitions alongside their corresponding classes. Perhaps the convention could be:

```rb
# variable_lookup.rb
module Liquid
  class VariableLookup
    class Traversal < Liquid::Traversal
      def children
        @node.lookups
      end
    end
  end
end
```

or do away with this abstraction altogether and put a children method on VariableLookup etc directly. If the latter, picking the method name is the hardest problem, given the rant about type consistency I posted earlier. Perhaps something to the effect of `children_of_all_types`, maybe phrased more eloquently. Then the new public accessors can be removed as well. Reflecting back, I think my dislike of the method added in the original PR was:

- the method name was ambiguous
- I'd rather have consumers use the visitor pattern to access the data

Having the method at all wasn't the issue for me.